Android/app/src/main/java/app/intra/IntroDialog.java

```diff
@@ -0,0 +1,187 @@
+  private static boolean shown = false;
+
+  public static boolean shouldShow(Activity activity) {
+    return !shown && !PersistentState.getWelcomeApproved(activity);
+  }
+
+  @Override
+  public int getTheme() {
+    return R.style.FullScreenDialog;
+  }
+
+  @Override
+  public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle
+      savedInstanceState) {
+    View welcomeView = inflater.inflate(R.layout.intro_pager, container);
+
+    Adapter adapter = new Adapter(getChildFragmentManager());
+    ViewPager pager = welcomeView.findViewById(R.id.welcome_pager);
+    pager.setAdapter(adapter);
+
+    // Right-to-left locale support.
+    boolean rtl = false;
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
+      rtl = welcomeView.getLayoutDirection() == View.LAYOUT_DIRECTION_RTL;
+    }
+
+    Button backButton = welcomeView.findViewById(R.id.intro_back);
+    // Workaround for lack of vector drawable support pre-Lollipop with drawableEnd.
+    // Using a bitmap drawable might be a simpler solution, but bitmaps don't have auto-mirroring.
+    // ic_chevron_start is auto-mirroring, so we just have to put it on the outside end.
+    Drawable startChevron = getResources().getDrawable(R.drawable.ic_chevron_start);
+    if (rtl) {
+      backButton.setCompoundDrawablesWithIntrinsicBounds(null, null, startChevron, null);
+    } else {
+      backButton.setCompoundDrawablesWithIntrinsicBounds(startChevron, null, null, null);
+    }
+    backButton.setOnClickListener(new View.OnClickListener() {
+      @Override
+      public void onClick(View view) {
+        int currentItem = pager.getCurrentItem();
+        if (currentItem > 0) {
+          pager.setCurrentItem(currentItem - 1);
+        }
+      }
+    });
+
+    Button nextButton = welcomeView.findViewById(R.id.intro_next);
+    Drawable endChevron = getResources().getDrawable(R.drawable.ic_chevron_end);
+    if (rtl) {
+      nextButton.setCompoundDrawablesWithIntrinsicBounds(endChevron, null, null, null);
+    } else {
+      nextButton.setCompoundDrawablesWithIntrinsicBounds(null, null, endChevron, null);
+    }
+    nextButton.setOnClickListener(new View.OnClickListener() {
+      @Override
+      public void onClick(View view) {
+        int currentItem = pager.getCurrentItem();
+        if (currentItem < NUM_PAGES - 1) {
+          pager.setCurrentItem(currentItem + 1);
+        }
+      }
+    });
+
+    Button acceptButton = welcomeView.findViewById(R.id.intro_accept);
+    acceptButton.setOnClickListener(new View.OnClickListener() {
+      @Override
+      public void onClick(View view) {
+        PersistentState.setWelcomeApproved(getContext(), true);
+        DnsVpnController.getInstance().start(getContext());
+        dismiss();
+      }
+    });
+
+    pager.addOnPageChangeListener(new TabListener(backButton, nextButton, acceptButton));
+
+    shown = true;
+    return welcomeView;
+  }
+
+  public static class Page0 extends Fragment {
```

Comment:
These class names and duplication seem like a smell to me.
Could we have a single class for the three pages? We could expose a property to set the one different string and in the case of `Page2`, a method to set the `details`?

Reply:
OK, I've unified the pages into a single layout and fragment class.  PTAL.